### PR TITLE
[PORT] Custom AI Holographic Appearance

### PR DESCRIPTION
**Pull request by Bobbanz1**
## About The Pull Request
This PR does the following:
1. Gives the AI players the ability to utilize their own player characters as a custom hologram.
2. Shifts carbon update_transform to living to reduce copypaste in weird edge cases.
## Ports From: Citadel Station 13
* https://github.com/Citadel-Station-13/Citadel-Station-13/pull/10191 (Half of it)

## Why It's Good For The Game
~~SNOWFLAKES~~
Gives AI holograms some (1) more lovely option to pick from by giving them the ability to pick their original character as a Hologram, now you too can roleplay as a brain in a box by using the original body as a hologram...kinda messed up if ya think about it.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

https://user-images.githubusercontent.com/59128051/226203343-a7bb7460-deb0-4b08-aba6-0bdec5b86452.mp4

</details>

## Changelog
:cl:Bobbanz1, silicons
add: Custom holoforms have been added for AIs. 
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/update_icons.dm [hunk 1]
```diff
@@ -1,28 +1,4 @@
...
-					final_dir = pick(NORTH, SOUTH) //So you fall on your side rather than your face or ass
-	if(resize != RESIZE_DEFAULT_SIZE)
-		changed++
-		ntransform.Scale(resize)
-		resize = RESIZE_DEFAULT_SIZE
-
-	if(changed)
-		animate(src, transform = ntransform, time = (lying_prev == 0 || !lying) ? 2 : 0, pixel_y = final_pixel_y, dir = final_dir, easing = (EASE_IN|EASE_OUT))
-		setMovetype(movement_type & ~FLOATING)  // If we were without gravity, the bouncing animation got stopped, so we make sure we restart it in next life().
+//NSV13 - Moved update_transform() to another file
```
Comment (covertcorvid) [inline code/modules/mob/living/carbon/update_icons.dm:1]: Why was this moved?

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 2]
```diff
@@ -829,18 +829,26 @@
...
+							holo_icon = client.prefs.get_filtered_holoform(HOLOFORM_FILTER_AI)
+						else
+							holo_icon = getHologramIcon(icon('icons/mob/ai.dmi', "default"))
 					if("xeno queen")
+					//NSV13 - AI Custom Holographic Form - Stop
 						holo_icon = getHologramIcon(icon(icon_list[input],"alienq"))
 					else
 						holo_icon = getHologramIcon(icon(icon_list[input], input))
-	return
+	//return //NSV13 - Disabled this because the port demands it
```
Comment (covertcorvid) [inline code/modules/mob/living/silicon/ai/ai.dm:851]: What difference did this make? Isn't it still hitting the end of the function?

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 3]
```diff
@@ -829,18 +829,26 @@
...
 				qdel(holo_icon)
 				switch(input)
+					//NSV13 - AI Custom Holographic Form - Start
+					if("custom")
+						if(client?.prefs?.custom_holoform_icon)
+							holo_icon = client.prefs.get_filtered_holoform(HOLOFORM_FILTER_AI)
+						else
+							holo_icon = getHologramIcon(icon('icons/mob/ai.dmi', "default"))
 					if("xeno queen")
+					//NSV13 - AI Custom Holographic Form - Stop
```
Comment (covertcorvid) [inline code/modules/mob/living/silicon/ai/ai.dm:847]: Can probably move one line up

File: code/modules/mob/living/silicon/ai/custom_holoform.dm [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/mob/living/silicon/ai/verb/clear_custom_holoform()
```
Comment (covertcorvid) [inline code/modules/mob/living/silicon/ai/custom_holoform.dm:1]: Since this file is NSV-only, please put it in the NSV directory

File: code/modules/mob/living/silicon/ai/freelook/eye.dm [hunk 5]
```diff
@@ -73,7 +73,7 @@
 // Use this when setting the ai_eye's location.
 // It will also stream the chunk that the new loc is in.
 
-/mob/camera/ai_eye/proc/setLoc(destination, force_update = FALSE)
+/mob/camera/ai_eye/proc/setLoc(destination, force_update = FALSE, dir) //NSV13 - AI Custom Holographic Form
```
Comment (covertcorvid) [inline code/modules/mob/living/silicon/ai/freelook/eye.dm:76]: I'm surprised the linter didn't complain about this. Usually unnamed parameters are required to come before named ones.

**Review by covertcorvid — APPROVED** [review]
Nothing glaring, the named/unnamed parameters is the most likely to cause issues

File: code/modules/mob/living/carbon/update_icons.dm (hunk 1 above)
Comment (Bobbanz1) [reply code/modules/mob/living/carbon/update_icons.dm:1]: I was following the changes made from the other PR thinking it was going to be needed somewhere, after having looked it through, it appears this was only moved because the pAI thing that I didn't port apparently needed it. 

Next Commit dealing with the reviews will move it back.

**Review by covertcorvid — DISMISSED** [review]

File: code/modules/mob/living/silicon/ai/freelook/eye.dm (hunk 5 above)
Comment (Bokkiewokkie) [reply code/modules/mob/living/silicon/ai/freelook/eye.dm:76]: They're all named because it's a proc definition, there's no restrictions on the position of predefined entries in a definition though.

File: code/__HELPERS/custom_holoforms.dm [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+			to_chat(user, "<span class='boldwarning'>You are attempting to set your custom holoform too fast!</span>")
+			return
+	return generate_custom_holoform_from_prefs(prefs, null, null, TRUE, TRUE)
+
+//Prompts this client for custom holoform parameters.
+/proc/user_interface_custom_holoform(client/C)
+	var/datum/preferences/target_prefs = C.prefs
+	ASSERT(target_prefs)
+	//In the future, maybe add custom path allowances a la admin create outfit but for now..
+	return generate_custom_holoform_from_prefs_safe(target_prefs, C.mob)
```
Comment (Bokkiewokkie) [inline code/__HELPERS/custom_holoforms.dm:58]: This file can also be moved to the NSV folder

File: code/__DEFINES/misc.dm [hunk 7]
```diff
@@ -471,3 +472,10 @@ GLOBAL_LIST_EMPTY(monkey_icon_cache)
 // Pirates threat
 #define PIRATE_RESPONSE_NO_PAY "pirate_answer_no_pay"
 #define PIRATE_RESPONSE_PAY "pirate_answer_pay"
+
+//NSV13 - AI Custom Holographic Form - Start
+#define CUSTOM_HOLOFORM_DELAY 10 SECONDS //prevents spamming to make lag. it's pretty expensive to do this.
+
+#define HOLOFORM_FILTER_AI "FILTER_AI"
+#define HOLOFORM_FILTER_STATIC "FILTER_STATIC"
+//NSV13 - AI Custom Holographic Form - Stop
```
Comment (Bokkiewokkie) [inline code/__DEFINES/misc.dm:481]: Defines can be moved to the NSV defines as well

**Review by Bokkiewokkie — DISMISSED** [review]
Just some more things we can move, otherwise it all looks god

File: code/__DEFINES/misc.dm (hunk 7 above)
Comment (Bobbanz1) [reply code/__DEFINES/misc.dm:481]: Can't move those as it produces a compile error if moved to an NSV file

Comment (Bobbanz1) [reply code/__DEFINES/misc.dm:481]: Never mind

**Review by covertcorvid — APPROVED** [review]

**Review by Bokkiewokkie — APPROVED** [review]
